### ISSUE TITLE
buildkitd Crashes with `CGroups V2 - sh: write error: Not supported`

### ISSUE THREAD

**Issue by xgqt** [earthly/earthly#2838]
Hello! 

Im trying to package earthly for Gentoo Linux and as a test project I'm using https://github.com/earthly/hello-world.

When running `earthly +hello` I get the following error:
```
 Init 🚀————————————————————————————————————————————————————————————————————————————————

           buildkitd | Starting buildkit daemon as a docker container (earthly-buildkitd)...
Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
Error: build new buildkitd client: maybe start buildkitd: wait until started: buildkitd crashed
It seems that buildkitd is shutting down or it has crashed. You can report crashes at https://github.com/earthly/earthly/issues/new.

================================= System Info ==================================

version: 0.7.2
build-sha: e6fed708700425fc2616bd1e7a0f202d414f5f23
platform: linux/amd64; Gentoo 

================================ Docker Version ================================

&containerutil.FrontendInfo{ClientVersion:"20.10.22", ClientAPIVersion:"1.41", ClientPlatform:"linux/amd64", ServerVersion:"20.10.22", ServerAPIVersion:"1.41", ServerPlatform:"linux/amd64", ServerAddress:"/var/run/docker.sock"}

================================ Buildkit Logs =================================

starting earthly-buildkit with EARTHLY_GIT_HASH=e6fed708700425fc2616bd1e7a0f202d414f5f23 BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:f8a515367d7b629d375897715ff28ea5ad9c0c8f+build
detected cgroups v2; buildkit/entrypoint.sh running under pid=1

```

The earthly-buildkitd logs:
```
starting earthly-buildkit with EARTHLY_GIT_HASH=e6fed708700425fc2616bd1e7a0f202d414f5f23 BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:f8a515367d7b629d375897715ff28ea5ad9c0c8f+build
detected cgroups v2; buildkit/entrypoint.sh running under pid=1
sh: write error: Not supported
```

`sh: write error: Not supported` looks very suspicious, you you think there is something wrong with the docker installation or with buildkitd container?

**Comment by xgqt** [comment]
Running latest buildkitd with debugging shows:
```
starting earthly-buildkit with EARTHLY_GIT_HASH=e6fed708700425fc2616bd1e7a0f202d414f5f23 BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:f8a515367d7b629d375897715ff28ea5ad9c0c8f+build
+ '[' -z 0 ]
+ '[' -z 0 ]
+ '[' -z true ]
+ '[' -z 20 ]
+ '[' -z /tmp/earthly ]
+ '[' -z cni ]
+ '[' -z 2 ]
+ '[' -f /sys/fs/cgroup/cgroup.controllers ]
+ echo 'detected cgroups v2; buildkit/entrypoint.sh running under pid=1'
+ mkdir -p /sys/fs/cgroup/earthly
detected cgroups v2; buildkit/entrypoint.sh running under pid=1
+ mkdir -p /sys/fs/cgroup/buildkit
+ echo 1
sh: write error: Not supported
```

The directories /sys/fs/cgroup/earthly and /sys/fs/cgroup/buildkit do not exist.

**Comment by jriedel-osp** [comment]
I had a similar issue on Ubuntu. After some frustration I was able to solve it by changing the `cgroupdriver` for docker (from `cgroupfs`) to `systemd`.   
Sadly I don't really know why that worked, but this was the only thing that had done it for me.    
Maybe this helps with further investigation.

**Comment by stellarpower** [comment]
That's interesting as I'm using podman and systemd should be the default cgroups driver anyway.

**Comment by stellarpower** [comment]
Would it also make sense to rename this issue to e.g. _"buildkitd Crashes with CGroups V2 - `sh: write error: Not supported`"_ ?

**Comment by alexcb** [comment]
maybe https://github.com/earthly/earthly/blob/main/buildkitd/entrypoint.sh#L45-L63 needs to be adjusted for `cgroupfs` support?

**Comment by alexcb** [comment]
> Slight aside, but is it possible to get the entrypoint to throw out a line number when it fails?

```
if [ "$BUILDKIT_DEBUG" = "true" ]; then
    set -x
fi
```

you'll need to define that; https://github.com/earthly/earthly/blob/main/CONTRIBUTING.md#running-buildkit-under-debug-mode shows an example for defining a different env.

**Comment by Grunet** [comment]
Running into this while trying Earthly for the first time. Using Podman inside of a Github Codespace running an Alpine dev container. 

Starting a Codespace from the https://github.com/Grunet/cloud-native-honeypot/tree/install-earthly branch and trying to do `sudo earthly --verbose bootstrap` or `sudo earthly --verbose github.com/earthly/hello-world+hello` seem to reproduce it.

I currently have it set to Earthly version 0.7.11 in https://github.com/Grunet/cloud-native-honeypot/blob/install-earthly/.devcontainer/Dockerfile

**Comment by stellarpower** [comment]
> you'll need to define that; https://github.com/earthly/earthly/blob/main/CONTRIBUTING.md#running-buildkit-under-debug-mode shows an example for defining a different env.

Sorry for the delay, so I get this debug output:
```
starting earthly-buildkit with EARTHLY_GIT_HASH=4ee9f89698b37f680ae0062191f66e20c4dd997c BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:e0f2b80558fa91b090bf9f489e033e995ca1106a+build
+ '[' -z 0 ]
+ '[' -z 0 ]
+ '[' -z true ]
+ '[' -z 20 ]
+ '[' -z /tmp/earthly ]
+ '[' -z cni ]
+ '[' -z 2 ]
+ '[' -f /sys/fs/cgroup/cgroup.controllers ]
+ echo 'detected cgroups v2; buildkit/entrypoint.sh running under pid=2'
detected cgroups v2; buildkit/entrypoint.sh running under pid=2
+ mkdir -p /sys/fs/cgroup/earthly
+ mkdir -p /sys/fs/cgroup/buildkit
+ echo 2
+ wc -l
+ '[' 7 '!=' 0 ]
+ echo 'warning: processes exist in the root cgroup; this may cause errors during cgroup initialization'
warning: processes exist in the root cgroup; this may cause errors during cgroup initialization
+ sleep 2
+ echo +pids
sh: write error: Resource busy
```

It's failing [here](https://github.com/earthly/earthly/blob/v0.7.11/buildkitd/entrypoint.sh#L55). Interestingly, if I just start the container using the CLI that running earthly with --verbose spits out, I can write to that file okay; but running the entrypoint script it complains about the write error. 

I also notice that the number of processes in the cgroup goes up by one each time I run. I haven;t looked into what buildkit is doing or how, but, is it possibly due to previous failed starts that have left lingering PIDs in the group, or some other stuff that hasn't been cleaned up appropriately?

When I upgrade earthly, or have had errors output, for example, I stop the buildkit container and then remove it to try and get a clean start. Is it possible that doing this has prevented some cleanup from happening (RAII for the win again!!)?

Thanks

**Comment by alexcb** [comment]
> I also notice that the number of processes in the cgroup goes up by one each time I run

This sounds very odd.

I think looking into why you are getting `warning: processes exist in the root cgroup; this may cause errors during cgroup initialization` would be helpful.

Can you `cat /sys/fs/cgroup/cgroup.procs` (and looking the process names, not just pids) to see what the processes remain in the root cgroup? Having processes here is likely what's causing the `subtree_control` updates to fail.

**Comment by alexcb** [comment]
in the container where https://github.com/earthly/earthly/blob/main/buildkitd/entrypoint.sh is being executed.

**Comment by stellarpower** [comment]
There are loads of duplicate lines with 0 - presume these are PIDs that are masked. Then it's just my shell session left, and I presume the few lingering IDs that come back with nothing are the ps command itself or something else running as part of the command pipeline I just ran (cat, sort, uniq, etc.)

**Comment by stellarpower** [comment]
Is there any other info I can provide? This has been blocking my being able to use earthly for a while, and I must admit I do miss it! I am trying out xmake a an improved C++ build system and it'd be lovely to be able to unite the two and hopefully cut down some of the pains. The way xmake manages caching too should help out a lot with Earthly integration.

**Comment by alexcb** [comment]
https://github.com/earthly/earthly/releases/tag/v0.7.14 has just been released, which contains a buildkit update which fixed some cgroup initialization and allowed us to remove our own cgroup entrypoint code -- can you give that a try to see if it helps?

**Comment by stellarpower** [comment]
Excellent, thank you! I will stress test i over the coming weeks but I can run the hello world example from the docs.

**Comment by alexcb** [comment]
if it does show up again please reopen this issue.

**Comment by xgqt** [comment]
This issue is solved for me with Earthly version 0.17.4. Tested it with the `hello` example and with a personal project. Thanks!
